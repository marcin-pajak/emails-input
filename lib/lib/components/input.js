"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeInputHandlers = exports.inputTemplate = exports.MIN_LENGTH = void 0;
var error_1 = require("../../utils/error");
var constants_1 = require("../constants");
exports.MIN_LENGTH = 21;
/**
 * Input Component Template
 */
exports.inputTemplate = function () {
    return "<input\n    aria-label=\"add more people\"\n    class=\"" + constants_1.CLASSES.INPUT + "\"\n    max-length=\"255\"\n    placeholder=\"add more people...\"\n  />";
};
/**
 * Input Component Event Handlers
 * @param elements
 * @param handlers
 */
exports.makeInputHandlers = function (elements, handlers) {
    var add = handlers.add;
    /**
     * Return input element
     */
    var getInput = function () {
        var input = elements.input;
        if (input === null) {
            throw new error_1.ErrorWithCode('Library is in wrong state.', error_1.ErrorCodes.ERR_INTERNAL);
        }
        return input;
    };
    /**
     * Add email on enter and comma
     * @param event
     */
    var handleInputChange = function (event) {
        var input = getInput();
        var key = event.key;
        var trimmed = input.value.trim();
        if (trimmed === '')
            return;
        var inputWidth = Math.max(input.value.length + 1, exports.MIN_LENGTH);
        input.style.minWidth = String(inputWidth) + "ch";
        if (key === 'Enter') {
            add(input.value);
            input.value = '';
        }
        else if (key === ',' && input.value.length > 1) {
            add(input.value.slice(0, Math.max(0, input.value.length - 1)));
            input.value = '';
        }
    };
    /**
     * Add email on paste
     * @param event
     */
    var handleInputPaste = function (event) {
        var _a, _b;
        event.preventDefault();
        event.stopImmediatePropagation();
        var input = getInput();
        var pastedText = (_b = ((_a = event.clipboardData) !== null && _a !== void 0 ? _a : window.clipboardData)) === null || _b === void 0 ? void 0 : _b.getData('text').trim();
        if (pastedText !== '' && pastedText !== undefined) {
            add(pastedText);
            input.value = '';
        }
        return false;
    };
    /**
     * Add email on blur
     */
    var handleInputBlur = function () {
        var input = getInput();
        var trimmedValue = input.value.trim();
        if (trimmedValue !== '') {
            add(trimmedValue);
            input.value = '';
        }
    };
    return {
        handleInputBlur: handleInputBlur,
        handleInputChange: handleInputChange,
        handleInputPaste: handleInputPaste,
    };
};
//# sourceMappingURL=input.js.map