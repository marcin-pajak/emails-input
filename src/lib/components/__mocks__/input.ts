export const inputTemplate = (): string => '<input />';
