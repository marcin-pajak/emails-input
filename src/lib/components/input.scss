.EmailsInput-input {
  background: #fff;
  border: none;
  box-sizing: border-box;
  color: #050038;
  font-size: 14px;
  height: 27px;
  line-height: 27px;
  flex-grow: 1;
  font-family: 'Open Sans', sans-serif;
  padding: 0;
  margin: 4px;
  outline: none;
  transition: color 0.3s ease-out;
}

.EmailsInput-input:focus,
.EmailsInput-input:hover {
  color: #000;
}

.EmailsInput-input::placeholder {
  color: #c3c2cf;
}
