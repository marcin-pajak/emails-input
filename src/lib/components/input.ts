import { ErrorCodes, ErrorWithCode } from '../../utils/error';
import { EmailActions } from '../actions/types';
import { CLASSES } from '../constants';
import { WindowWithClipboard } from '../types';
import { Elements, InputHandlers } from './types';

export const MIN_LENGTH = 21;

/**
 * Input Component Template
 */
export const inputTemplate = (): string =>
  `<input
    aria-label="add more people"
    class="${CLASSES.INPUT}"
    max-length="255"
    placeholder="add more people..."
  />`;

/**
 * Input Component Event Handlers
 * @param elements
 * @param handlers
 */
export const makeInputHandlers = (
  elements: Elements,
  handlers: EmailActions,
): InputHandlers => {
  const { add } = handlers;

  /**
   * Return input element
   */
  const getInput = (): HTMLInputElement => {
    const { input } = elements;

    if (input === null) {
      throw new ErrorWithCode(
        'Library is in wrong state.',
        ErrorCodes.ERR_INTERNAL,
      );
    }

    return input;
  };

  /**
   * Add email on enter and comma
   * @param event
   */
  const handleInputChange = (event: KeyboardEvent): void => {
    const input = getInput();
    const { key } = event;
    const trimmed = input.value.trim();

    if (trimmed === '') return;

    const inputWidth = Math.max(input.value.length + 1, MIN_LENGTH);
    input.style.minWidth = `${String(inputWidth)}ch`;

    if (key === 'Enter') {
      add(input.value);
      input.value = '';
    } else if (key === ',' && input.value.length > 1) {
      add(input.value.slice(0, Math.max(0, input.value.length - 1)));
      input.value = '';
    }
  };

  /**
   * Add email on paste
   * @param event
   */
  const handleInputPaste = (event: ClipboardEvent): boolean => {
    event.preventDefault();
    event.stopImmediatePropagation();

    const input = getInput();
    const pastedText = (
      event.clipboardData ??
      ((window as unknown) as WindowWithClipboard).clipboardData
    )
      ?.getData('text')
      .trim();

    if (pastedText !== '' && pastedText !== undefined) {
      add(pastedText);
      input.value = '';
    }

    return false;
  };

  /**
   * Add email on blur
   */
  const handleInputBlur = (): void => {
    const input = getInput();
    const trimmedValue = input.value.trim();

    if (trimmedValue !== '') {
      add(trimmedValue);
      input.value = '';
    }
  };

  return {
    handleInputBlur,
    handleInputChange,
    handleInputPaste,
  };
};
