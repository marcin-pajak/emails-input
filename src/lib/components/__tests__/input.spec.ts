import { fireEvent, screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';

import EmailsInput from '../../..';
import { inputTemplate } from '../input';

describe('input template should', (): void => {
  it('return valid template', (): void => {
    expect.hasAssertions();

    expect(inputTemplate()).toMatchSnapshot();
  });
});

const ELEMENT_CLASS = 'emails-input';
describe('input component should', (): void => {
  beforeEach((): void => {
    document.body.innerHTML = `<div id="${ELEMENT_CLASS}"></div>`;
    const element = document.querySelector(`#${ELEMENT_CLASS}`) as HTMLElement;
    EmailsInput(element);
  });

  afterEach((): void => {
    const element = document.querySelector(`#${ELEMENT_CLASS}`);
    element?.remove();
  });

  it('handle adding value after comma', async (): Promise<void> => {
    expect.hasAssertions();

    await userEvent.type(screen.getByRole('textbox'), 'some@email.com,');

    expect(await screen.findByText('some@email.com')).not.toBeNull();
  });

  it('handle adding value after enter', async (): Promise<void> => {
    expect.hasAssertions();

    await userEvent.type(screen.getByRole('textbox'), 'some@email.com{enter}');

    expect(await screen.findByText('some@email.com')).not.toBeNull();
  });

  it('handle adding value after blur', async (): Promise<void> => {
    expect.hasAssertions();

    const input = screen.getByRole('textbox');
    await userEvent.type(input, 'some@email.com');
    fireEvent.blur(input);

    expect(await screen.findByText('some@email.com')).not.toBeNull();
  });

  it('handle adding value paste', async (): Promise<void> => {
    expect.hasAssertions();

    // eslint-disable-next-line @typescript-eslint/dot-notation
    window['clipboardData'] = {
      getData: jest.fn((): string => 'some@email.com'),
    };
    const input = screen.getByRole('textbox');
    userEvent.paste(input, 'some@email.com');

    expect(await screen.findByText('some@email.com')).not.toBeNull();
  });

  it('not add value after enter', async (): Promise<void> => {
    expect.hasAssertions();

    await userEvent.type(screen.getByRole('textbox'), '  {enter}');

    expect(screen.queryAllByText('  ')).toHaveLength(0);
  });

  it('not add value after blur', async (): Promise<void> => {
    expect.hasAssertions();

    const input = screen.getByRole('textbox');
    await userEvent.type(input, '  ');
    fireEvent.blur(input);

    expect(screen.queryAllByText('  ')).toHaveLength(0);
  });
});
