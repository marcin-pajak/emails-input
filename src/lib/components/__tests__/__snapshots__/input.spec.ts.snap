// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input template should return valid template 1`] = `
"<input
    aria-label=\\"add more people\\"
    class=\\"EmailsInput-input\\"
    max-length=\\"255\\"
    placeholder=\\"add more people...\\"
  />"
`;
